File: go/coercion.go
// Coercion: lenient JSON decoding for Safaricom's inconsistent value types.

package mpesa

import (
	"encoding/json"
	"fmt"
	"math"
	"strconv"
	"strings"
)

// FlexString coerces JSON strings or numbers into a Go string. Daraja mixes
// both (e.g. STK Query ResultCode arrives as "1032" in some captures, 1032 in
// others).
type FlexString string

// UnmarshalJSON accepts quoted strings, bare numbers, booleans, and null.
func (f *FlexString) UnmarshalJSON(b []byte) error {
	s := strings.TrimSpace(string(b))
	if s == "null" || s == "" {
		*f = ""
		return nil
	}
	if len(s) >= 2 && s[0] == '"' && s[len(s)-1] == '"' {
		var v string
		if err := json.Unmarshal(b, &v); err != nil {
			return err
		}
		*f = FlexString(v)
		return nil
	}
	var num json.Number
	if err := json.Unmarshal(b, &num); err == nil {
		*f = FlexString(num.String())
		return nil
	}
	*f = FlexString(s)
	return nil
}

// String returns the coerced wire representation of the value.
func (f FlexString) String() string { return string(f) }

// FlexInt64 coerces JSON numbers or numeric strings into an int64. OAuth's
// expires_in arrives as the STRING "3599" in official captures and as a bare
// number elsewhere.
type FlexInt64 int64

// UnmarshalJSON accepts quoted numeric strings, bare numbers, and null;
// null or "" map to 0 — callers treat <=0 as TTL unknown. Malformed input
// (doubled quotes, alphabetic content) is a hard error. Integral floats
// like 3599.0 are accepted (Python coerce_int + TS safeJsonInt parity);
// non-integral floats like 1.5 are rejected.
func (f *FlexInt64) UnmarshalJSON(b []byte) error {
	s := strings.TrimSpace(string(b))
	if s == "null" || s == "" {
		*f = 0
		return nil
	}
	if len(s) >= 2 && s[0] == '"' && s[len(s)-1] == '"' {
		var str string
		if err := json.Unmarshal(b, &str); err != nil {
			return err
		}
		str = strings.TrimSpace(str)
		if str == "" {
			*f = 0
			return nil
		}
		// Strict integer fast path: plain ASCII digits with optional sign.
		if v, err := strconv.ParseInt(str, 10, 64); err == nil {
			*f = FlexInt64(v)
			return nil
		}
		// Lenient fallback: integral floats only ("3599.0" → 3599);
		// non-integral, non-finite or out-of-range values stay hard errors.
		// Pattern copied from classification.go parseResultCode.
		if v, err := parseIntegralFloat(str); err == nil {
			*f = FlexInt64(v)
			return nil
		}
		return fmt.Errorf("mpesa: cannot parse %s as integer", b)
	}
	// Strict integer fast path for bare numbers.
	if v, err := strconv.ParseInt(s, 10, 64); err == nil {
		*f = FlexInt64(v)
		return nil
	}
	// Lenient fallback: bare integral floats only (3599.0 → 3599).
	if v, err := parseIntegralFloat(s); err == nil {
		*f = FlexInt64(v)
		return nil
	}
	return fmt.Errorf("mpesa: cannot parse %s as integer", b)
}

// parseIntegralFloat converts an integral-float rendering to int64,
// rejecting non-integral fractions, non-finite values and out-of-int64
// magnitudes. It mirrors classification.go's lenient fallback so OAuth
// TTLs (Python coerce_int integral-float branch, TS safeJsonInt) accept
// 3599.0 while 1.5/NaN/Inf/1e30 stay errors.
func parseIntegralFloat(s string) (int64, error) {
	f, err := strconv.ParseFloat(s, 64)
	if err != nil || math.IsNaN(f) || math.IsInf(f, 0) || f != math.Trunc(f) || f < math.MinInt64 || f > math.MaxInt64 {
		return 0, fmt.Errorf("not an integral float: %q", s)
	}
	return int64(f), nil
}

File: typescript/src/coercion.ts
/**
 * Lenient coercion helpers for Safaricom's inconsistent JSON value types.
 *
 * Daraja sends the same logical field with different JSON types depending
 * on endpoint: `ResponseCode` arrives as the STRING "0" in synchronous STK
 * responses but as INTEGER 0 inside callbacks; `expires_in` is the STRING
 * "3599" in official OAuth captures and a bare number elsewhere (see
 * docs/apis/stk-push.md, docs/apis/oauth.md). Callback metadata such as
 * `TransactionDate` is numeric-typed (docs/apis/stk-push.md callback).
 *
 * **Callback data is JSON with string-ified numbers**: every function here
 * accepts every observed shape and canonicalizes. Numbers are guarded to the
 * ±2^53 safe-integer range; values beyond that boundary return null/fallback
 * because their integer conversion silently corrupts precision (e.g. `1e30`
 * carries ~19 spurious digits).
 *
 * @packageDocumentation
 */

/** INT32_MAX sentinel — Daraja signals 32-bit overflow with this exact value. */
const INT32_MAX = 2_147_483_647;

/**
 * Attempt to parse `value` as a JSON number via `JSON.parse(String(value))`.
 * Returns the result if it is a finite safe integer (±2^53); returns null on
 * failure, overflow, NaN, Infinity, or non-numeric JSON types. Never throws.
 *
 * @example
 * ```ts
 * safeJsonInt("3599");  // 3599
 * safeJsonInt(3599);    // 3599
 * safeJsonInt("abc");   // null
 * safeJsonInt("9007199254740993"); // null (beyond ±2^53)
 * ```
 */
export function safeJsonInt(value: unknown): number | null {
  try {
    const parsed = JSON.parse(String(value));
    if (typeof parsed !== "number") return null;
    return Number.isFinite(parsed) && Number.isSafeInteger(parsed) ? parsed : null;
  } catch {
    return null;
  }
}

/**
 * Fast ASCII-digit check: returns true if `value` is a string matching
 * `/^-?\d{1,15}$/` with no leading zeros on multi-digit numbers. Rejects
 * Unicode-ND digits (e.g. "٠٢٣"), commas, whitespace, and overflow-length
 * strings. Capped at 15 digits because any 15-digit integer fits in JS's
 * ±2^53 safe range.
 *
 * @example
 * ```ts
 * isNumericString("123");   // true
 * isNumericString("-1");    // true
 * isNumericString("01");    // false (leading zero)
 * isNumericString("٠٢٣");  // false (Unicode-ND)
 * isNumericString("1,000"); // false (comma)
 * ```
 */
export function isNumericString(value: unknown): boolean {
  if (typeof value !== "string") return false;
  if (!/^-?\d{1,15}$/.test(value)) return false;
  const abs = value[0] === "-" ? value.slice(1) : value;
  return abs === "0" || abs[0] !== "0";
}

/**
 * Coerce `value` to an integer. Uses {@link safeJsonInt} for parsing; falls
 * back to `opts.fallback` (default 0) on failure. Throws `RangeError` when
 * the parsed value equals the INT_MAX sentinel (2^31-1), which Daraja uses
 * to signal 32-bit overflow.
 *
 * @param value - Raw input from JSON payload (string, number, null, etc.).
 * @param opts.fallback - Value returned when parsing fails (default: 0).
 * @param opts.field - Field name for the diagnostic message on RangeError.
 * @throws {RangeError} When the value equals INT32_MAX (2^31-1).
 *
 * @example
 * ```ts
 * coerceInt("3599");                     // 3599
 * coerceInt("abc");                      // 0 (default fallback)
 * coerceInt("abc", { fallback: -1 });   // -1
 * coerceInt(2147483647, { field: "expires_in" }); // throws RangeError
 * ```
 */
export function coerceInt(
  value: unknown,
  opts: { fallback?: number; field?: string } = {},
): number {
  const n = safeJsonInt(value);
  if (n === null) return opts.fallback ?? 0;
  if (n === INT32_MAX) {
    // TS-specific: Go/Python don't throw here — their int types absorb the value.
    const field = opts.field ?? "value";
    throw new RangeError(`mpesa: ${field} exceeds 32-bit range (${n})`);
  }
  return n;
}

/**
 * Convenience alias for {@link safeJsonInt}. Accepts string, number, null
 * or undefined and returns the parsed integer or null.
 *
 * @example
 * ```ts
 * parseIntSafe("3599");  // 3599
 * parseIntSafe(null);    // null
 * ```
 */
export function parseIntSafe(value: unknown): number | null {
  return safeJsonInt(value);
}
